### PR TITLE
[jit] Add missing tensor properties

### LINKED ISSUE #30775: Keep Tensor Properties in JIT & Eager in Sync
## 🚀 Feature

Currently there are a number of tensor properties which we bind in python which are not shared with the JIT / C++ api. 

Some of these have been duplicated in JIT, but many are not supported. 
Either:
- Develop a test mechanism to keep them in sync
- Figure out a way to share properties via native declaration

Native declaration would likely be a two-step process:
 - add properties to JIT type system (or other solution)
- figure out codegen

The following properties are currently bound, many of which will throw if they are used in a scripted function.

```
  {"T", (getter)THPVariable_get_T, nullptr, nullptr, nullptr},
  {"_cdata", (getter)THPVariable_get_cdata, nullptr, nullptr, nullptr},
  {"_version", (getter)THPVariable_get_version, nullptr, nullptr, nullptr},
  {"grad_fn", (getter)THPVariable_get_grad_fn, nullptr, nullptr, nullptr},
  {"_grad_fn", (getter)THPVariable_get_grad_fn, (setter)THPVariable_set_grad_fn, nullptr, nullptr},
  {"is_leaf", (getter)THPVariable_is_leaf, nullptr, nullptr, nullptr},
  {"data", (getter)THPVariable_get_data, (setter)THPVariable_set_data, nullptr, nullptr},
  {"_grad", (getter)THPVariable_get_grad, (setter)THPVariable_set_grad, nullptr, nullptr}, // only for legacy reasons
  {"grad", (getter)THPVariable_get_grad, (setter)THPVariable_set_grad, nullptr, nullptr},
  {"_base", (getter)THPVariable_get_base, nullptr, nullptr, nullptr},
  {"volatile", (getter)THPVariable_get_volatile, (setter)THPVariable_set_volatile, nullptr, nullptr},
  {"output_nr", (getter)THPVariable_get_output_nr, nullptr, nullptr, nullptr},
  {"requires_grad", (getter)THPVariable_get_requires_grad, (setter)THPVariable_set_requires_grad, nullptr, nullptr},
  {"_backward_hooks", (getter)THPVariable_get_backwards_hooks, (setter)THPVariable_set_backwards_hooks, nullptr, nullptr},
  {"name", (getter)THPVariable_get_name, nullptr, nullptr, nullptr},
  {"shape", (getter)THPVariable_get_shape, nullptr, nullptr, nullptr},
  {"is_cuda", (getter)THPVariable_is_cuda, nullptr, nullptr, nullptr},
  {"is_sparse", (getter)THPVariable_is_sparse, nullptr, nullptr, nullptr},
  {"is_mkldnn", (getter)THPVariable_is_mkldnn, nullptr, nullptr, nullptr},
  {"is_quantized", (getter)THPVariable_is_quantized, nullptr, nullptr, nullptr},
  {"dtype", (getter)THPVariable_dtype, nullptr, nullptr, nullptr},
  {"layout", (getter)THPVariable_layout, nullptr, nullptr, nullptr},
  {"device", (getter)THPVariable_device, nullptr, nullptr, nullptr},
  {"ndim", (getter)THPVariable_get_ndim, nullptr, nullptr, nullptr},
```

### PR DESCRIPTION
**Pull request by driazati**
Fixes #30775

This adds TorchScript implementations (copied from `python_variable.cpp`) for the remainin `Tensor` properties that were missing from the jit, in addition to a test that ensures new properties will trigger a failure so we can decide whether we want to add them as well.

For `some_tensor`, adds:

* `some_tensor.T`
* `some_tensor.ndim`
* `some_tensor.is_leaf`
* `some_tensor.name`

Differential Revision: [D20153288](https://our.internmc.facebook.com/intern/diff/20153288/)

### CODE REVIEW

File: torch/csrc/jit/runtime/register_prim_ops.cpp [hunk 1]
```diff
@@ -580,6 +580,55 @@ RegisterOperators reg(
...
+         "prim::T(Tensor(a) a) -> Tensor(a)",
+         [](Stack& stack) {
+           at::Tensor a;
+           pop(stack, a);
+           push(stack, a.numpy_T());
+           return 0;
+         },
+         aliasAnalysisFromSchema()),
+     Operator(
+         "prim::output_nr(Tensor a) -> int",
```
Comment (eellison) [inline torch/csrc/jit/runtime/register_prim_ops.cpp]: this isn't a user facing function, i'm not sure it's worth adding.

File: torch/csrc/jit/runtime/register_prim_ops.cpp [hunk 2]
```diff
@@ -580,6 +580,55 @@ RegisterOperators reg(
...
+         [](Stack& stack) {
+           at::Tensor a;
+           pop(stack, a);
+           push(stack, a.output_nr());
+           return 0;
+         },
+         aliasAnalysisFromSchema()),
+     Operator(
+         "prim::ndim(Tensor a) -> int",
```
Comment (eellison) [inline torch/csrc/jit/runtime/register_prim_ops.cpp]: This is the same op which already exists as `aten::dim`, and which we have optimizations for https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/peephole.cpp#L363.  It would be nice if passes did not have to reason about duplicate passes & we didn't register an additional op that we then couldn't remove because of BC. 

Could you add a `ndim` function [here](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/frontend/builtin_functions.cpp#L40) under the `prim` namespace?
```
def ndim(x):
    return x.dim()
```

**Review by eellison — COMMENTED** [review]
Nice, this will be a good increase in parity. One comment about duplication of ops. Another solution could be to run a canonicalization pass, but then we still have the extra `prim::ndim` op for no good reason.

File: torch/csrc/jit/runtime/register_prim_ops.cpp [hunk 3]
```diff
@@ -580,6 +580,55 @@ RegisterOperators reg(
...
+         [](Stack& stack) {
+           at::Tensor a;
+           pop(stack, a);
+           push(stack, a.is_leaf());
+           return 0;
+         },
+         aliasAnalysisFromSchema()),
+     Operator(
+         "prim::T(Tensor(a) a) -> Tensor(a)",
+         [](Stack& stack) {
```
Comment (eellison) [inline torch/csrc/jit/runtime/register_prim_ops.cpp]: similarly, this could be registered in builtins without adding another op.
```
def T(x):
    return x.numpy_T()
```

File: torch/csrc/jit/runtime/register_prim_ops.cpp [hunk 4]
```diff
@@ -580,6 +580,28 @@ RegisterOperators reg(
            return 0;
          },
          aliasAnalysisFromSchema()),
+     Operator(
+         "prim::is_leaf(Tensor a) -> bool",
```
Comment (eellison) [inline torch/csrc/jit/runtime/register_prim_ops.cpp]: Looks like `is_leaf` also is natively registered, so it can go with the others.

**Review by eellison — APPROVED** [review]
🚀
